src/offchainapi/storage.py

```diff
@@ -231,20 +206,23 @@ def atomic_writes(self):
         return self
 
     def __enter__(self):
-        self.rlock.acquire()
         if self.levels == 0:
             self.current_transaction = get_unique_string()
+            self.db.set_write(True)
 
         self.levels += 1
 
-    def __exit__(self, type, value, traceback):
-        try:
-            self.levels -= 1
-            if self.levels == 0:
-                self.current_transaction = None
-                self.persist_cache()
-        finally:
-            self.rlock.release()
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.levels -= 1
+        if self.levels == 0:
+            self.current_transaction = None
```

Comment:
I'm thinking how this works in a concurrent scenario:
say it's processing request A and B concurrently, and they enter/exit the context manger in the following order (T1< T2 < T3 < T4):
T1: A enters
T2: B enters
T3: A exits, success (no exceptions), but it does not commit because level != 0 
T4: B exits, failure (exception is thrown), now the whole transaction is aborted, including A's successful execution